File: 프로그래머스/LV1/소수 찾기.py
import math

def solution(n):
    prime_nums = set(range(2,n+1))
    
    for i in range(2, int(math.sqrt(n))+1):
        if i in prime_nums:
            prime_nums -= set(range(i*2,n+1,i))
            
    return len(prime_nums)

File: 프로그래머스/javascript/LV2/소수 찾기.js
function solution(numbers) {
  const digits = numbers.split("");
  const visited = Array(digits.length).fill(false);
  const set = new Set();

  function dfs(current) {
    if (current !== "") {
      set.add(Number(current));
    }

    for (let i = 0; i < digits.length; i++) {
      if (visited[i]) continue;

      visited[i] = true;
      dfs(current + digits[i]);
      visited[i] = false;
    }
  }

  dfs("");

  let count = 0;
  for (let num of set) {
    if (isSosu(num)) count++;
  }

  return count;
}

function isSosu(number) {
  if (number < 2) return false;

  for (let i = 2; i <= Math.sqrt(number); i++) {
    if (number % i === 0) return false;
  }

  return true;
}
